### ISSUE TITLE
New control curve needs a name

### ISSUE THREAD

**Issue by snorfalorpagus** [pywr/pywr#357]
I've been using this control curve in a model to good effect. It has a fixed cost above the curve, and a linearly interpolated cost between two values (not related to the above curve value) when below the curve.

Suggestions for a parameter name?

```
from pywr.parameters import Parameter
class NewControlCurve(Parameter):
    def __init__(self, storage_node, control_curve, values, *args, **kwargs):
        super(NewControlCurve, self).__init__(*args, **kwargs)
        self.storage_node = storage_node
        self.control_curve = control_curve
        self.values = values

    def before(self, timestamp):
        self.current_pc = self.storage_node.current_pc[0]
    
    def value(self, timestamp, scenario_index):
        cc = self.control_curve.value(timestamp, scenario_index)
        if self.current_pc > cc:
            return self.values[1]
        else:
            a, b = self.values[0]
            x = self.current_pc / cc
            z = a + (b-a)*x
            return z
```

**Comment by jetuk** [comment]
Couldn't you use the normal CC but with a parameter in the second slot that returns value based on Storage?

**Comment by snorfalorpagus** [comment]
I hadn't thought of that. It's possible, although the syntax feels a bit verbose:

```{
    "type": "controlcurve",
    "storage_node": "MyReservoir",
    "control_curves": ["MyCC"],
    "parameters": [
        {
            "type": "constant",
            "value": 0
        },
        {
            "type": "controlcurveinterpolated",
            "storage_node": "MyReservoir",
            "control_curves": ["MyCC"],
            "values": [-100, -200]
        }
    ]
}
```

**Comment by jetuk** [comment]
Can you do it with the normal interpolated control but use the same values in the "upper" part?

**Comment by jetuk** [comment]
Sorry I see why you can't.

**Comment by snorfalorpagus** [comment]
Note that the code I posted isn't scenario aware...

**Comment by jetuk** [comment]
Scenario aware code below. I've not tested this. Should be cythonised really.

```
from pywr.parameters import Parameter
class NewControlCurve(Parameter):
    def __init__(self, storage_node, control_curve, values, *args, **kwargs):
        super(NewControlCurve, self).__init__(*args, **kwargs)
        self.storage_node = storage_node
        self.control_curve = control_curve
        self.values = values
    
    def value(self, timestamp, scenario_index):
        cc = self.control_curve.value(timestamp, scenario_index)
        pc = self.storage_node.current_pc[scenario_index.global_id]
        if pc > cc:
            return self.values[1]
        else:
            a, b = self.values[0]
            x = pc / cc
            z = a + (b-a)*x
            return z
```

**Comment by snorfalorpagus** [comment]
We should check for a divide by zero error here - it's not unreasonable to have a control curve at zero.

**Comment by jetuk** [comment]
The function is a bit weird actually. Is `a` usually zero? The `x` parameter is an inverse function, but the if condition truncates it. It's like a semi-inverse or something.

**Comment by jetuk** [comment]
Thinking about this... a related one would be to use the exponential or natural log function instead.

**Comment by snorfalorpagus** [comment]
I've been using it as (or similar to) the parameters in the JSON example above, i.e. 0 above the curve, then a jump to -100 below the curve, increasing to -200 as storage approaches zero.

**Comment by jetuk** [comment]
Getting a bit off topic now, but a parameter based on the generalised logistic function might be a good approach for balancing multiple storages.

https://en.m.wikipedia.org/wiki/Generalised_logistic_function

**Comment by snorfalorpagus** [comment]
Here's the most recent version of this:

```
from pywr.parameters import Parameter, load_parameter

class NewControlCurve(Parameter):
    def __init__(self, model, storage_node, control_curve, values, minimum=0.0, *args, **kwargs):
        super(NewControlCurve, self).__init__(model, *args, **kwargs)
        self.storage_node = storage_node
        self.control_curve = control_curve
        self.values = values # [(below_lower, below_upper), above]
        self.minimum = minimum # lower level for interpolation
        assert(len(values) == 2)
        assert(len(values[0]) == 2)
        assert(isinstance(values[1], (int, float)))
        assert(isinstance(values[0][0], (int, float)))
        self.children.add(self.control_curve)

    def value(self, timestamp, scenario_index):
        cc = self.control_curve.get_value(scenario_index)
        current_pc = self.storage_node.current_pc[scenario_index.global_id]
        if current_pc > cc:
            a, b = [self.values[1], self.values[1] + 1]  # TODO: this is a hack
            if cc == 1.0:
                x = 0.0
            else:
                x = (current_pc - cc) / (1.0 - cc)
            z = a + (b-a)*x
        else:
            a, b = self.values[0]
            x = (current_pc - self.minimum) / (cc - self.minimum)
            z = a + (b-a)*x
        return z

    @classmethod
    def load(cls, model, data):
        storage_node = model._get_node_from_ref(model, data["storage_node"])
        control_curve = load_parameter(model, data["control_curve"])
        values = data["values"]
        minimum = data.get("minimum", 0.0)
        parameter = NewControlCurve(model, storage_node, control_curve, values, minimum)
        return parameter

NewControlCurve.register()
```

**Comment by snorfalorpagus** [comment]
"PiecewiseLinearControlCurve"

**Comment by jetuk** [comment]
Solved in #566 and cythonised in #571
